tamr_unify_client/models/base_collection.py

```diff
@@ -69,3 +69,33 @@ def stream(self, resource_class):
 
     def __iter__(self):
         return self.stream()
+
+    @abstractmethod
+    def by_external_id(self, resource_class, external_id):
+        """Retrieve an item in this collection by relative ID.
+
+        Subclasses should override this method and pass in the specific
+        ``resource_class`` for that collection.
+
+        :param resource_class: Resource class corresponding to items in this collection.
+        :type resource_class: Python class
+        :param external_id: The external ID.
+        :type external_id: str
+        :returns: The specified item, if found.
+        :rtype: ``resource_class``
+        """
+        params = {"filter": "externalId==" + external_id}
+        resources = self.client.get(self.api_path, params=params).successful().json()
+        items = [
+            resource_class.from_json(self.client, resource_json)
+            for resource_json in resources
+        ]
+
+        if len(items) == 0:
+            raise KeyError(f'No item found with external ID "{external_id}"')
+        elif len(items) > 1:
+            raise ValueError(
```

Comment:
My gut instinct is that this should also be a `KeyError`. Usually `ValueError` is for telling the user that the params they passed in were not of the right type / format, whereas `KeyError` indicates that a lookup attempt failed.

Do you think `ValueError` is better than `KeyError` here?

Reply:
I picked `ValueError` because the error is that there are too many values, not that something was wrong with the key.  It's also something that shouldn't ever happen, because if it does it means the index in PG isn't enforcing uniqueness.

Reply:
let me know what you think; if I should change the second one to be `KeyError` (or something else), or if `ValueError` makes sense.

Reply:
TIL there's a built-in [`LookupError`](https://docs.python.org/3/library/exceptions.html#LookupError) in Python.

> The base class for the exceptions that are raised when a key or index used on a mapping or sequence is invalid: IndexError, KeyError. This can be raised directly by codecs.lookup().

I think we should use `LookupError` here

Reply:
for both?

Reply:
No, `KeyError` for the 0 case is exactly what `KeyError` is designed for so I think being specific there is good. Just changing the `ValueError` to `LookupError` is what I propose.

Reply:
done